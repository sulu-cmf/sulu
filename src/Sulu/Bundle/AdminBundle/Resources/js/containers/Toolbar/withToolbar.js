// @flow
import {autorun} from 'mobx';
import React from 'react';
import type {ComponentType} from 'react';
import {buildHocDisplayName} from '../../services/react';
import type {Item} from './types';
import toolbarStore from './stores/ToolbarStore';

export default function withToolbar(Component: ComponentType<*>, toolbar: () => Array<Item>) {
    const WithToolbarComponent = class extends React.Component<*> {
        disposer: Function;

        setToolbarItems = (component: React$Element<*>) => {
            this.disposer = autorun(() => {
                if (component) {
                    toolbarStore.setItems(toolbar.call(component));
                }
            });
        };

        componentWillUnmount() {
            this.disposer();
        }

        render() {
            return <Component ref={this.setToolbarItems} {...this.props} />;
        }
    };

    WithToolbarComponent.displayName = buildHocDisplayName('withToolbar', Component);

    return WithToolbarComponent;
}
