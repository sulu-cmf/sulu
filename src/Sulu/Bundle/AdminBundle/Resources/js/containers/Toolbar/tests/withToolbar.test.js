/* eslint-disable flowtype/require-valid-file-annotation */
import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import toolbarStore from '../stores/ToolbarStore';
import withToolbar from '../withToolbar';

jest.mock('../stores/ToolbarStore', () => {
    return {};
});

test('Pass props to rendered component', () => {
    toolbarStore.setItems = jest.fn();

    const Component = (props) => (<h1>{props.title}</h1>);
    const ComponentWithToolbar = withToolbar(Component, () => []);

    expect(ReactTestRenderer.create(<ComponentWithToolbar title="Test" />)).toMatchSnapshot();
});

test('Bind toolbar method to component instance', () => {
    toolbarStore.setItems = jest.fn();

    const Component = class Component extends React.Component {
        test = true;

        render() {
            return <h1>Test</h1>;
        }
    };

    const ComponentWithToolbar = withToolbar(Component, function() {
        return [{
            title: 'Save',
            icon: 'save',
            enabled: this.test,
        }];
    });

    expect(ReactTestRenderer.create(<ComponentWithToolbar />)).toMatchSnapshot();

    expect(toolbarStore.setItems).toBeCalledWith([{
        title: 'Save',
        icon: 'save',
        enabled: true,
    }]);
});
