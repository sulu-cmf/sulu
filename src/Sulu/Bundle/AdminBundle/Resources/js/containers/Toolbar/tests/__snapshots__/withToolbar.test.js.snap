// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bind toolbar method to component instance 1`] = `
<h1>
  Test
</h1>
`;

exports[`Pass props to rendered component 1`] = `
<h1>
  Test
</h1>
`;
